orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -223,7 +223,21 @@ Status GradientGraphBuilder::Build() {
     }
   }
 
-  return GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs);
+  ORT_RETURN_IF_ERROR(GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs));
+
+  std::unordered_map<std::string, int> node_index_to_node_name_map;
+  for (auto& node : graph_->Nodes()) {
+    node_index_to_node_name_map[node.Name()] = node.Index();
+  }
+  // Make sure the topological positions of backward nodes are after loss.
+  auto loss_node = graph_->GetProducerNode(loss_node_arg_name_);
+  for (auto& backward_node_def : gradient_graph_defs.NodeDefs()) {
+    const auto backward_node_name = backward_node_def.name;
+    const auto backward_node_index = node_index_to_node_name_map.at(backward_node_name);
+    ORT_ENFORCE(graph_->AddControlEdge(loss_node->Index(), backward_node_index));
```

Comment:
>AddControlEdge [](start = 24, length = 14)

This control edge thing is not quite working...BTW, why need to make sure backward nodes are after loss? In last stage we don't need events between FW/BW right?

Reply:
It's an anchor point for creating "loss -> Record forward Compute -> Wait backward Recv`.

---
In reply to: [476905060](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r476905060) [](ancestors = 476905060)

Reply:
from the code in pipeline_transformer.cc, line 500ish, it seems in last stage no record_forward_send nor wait_backward_recv

---
In reply to: [477034463](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r477034463) [](ancestors = 477034463,476905060)

Reply:
Sorry. It's only for making `loss -> Record forward Compute`.

---
In reply to: [477496037](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r477496037) [](ancestors = 477496037,477034463,476905060)

Reply:
by entire graph, do you mean the entire subgraph in the last stage? I think that would be OK..

---
In reply to: [477541868](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r477541868) [](ancestors = 477541868,477496037,477034463,476905060)

Reply:
It's not ok. There will be a big bubble because last-stage-FW becomes much much larger than last-stage-BW.

---
In reply to: [477549484](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r477549484) [](ancestors = 477549484,477541868,477496037,477034463,476905060)

Reply:
Forward-compute-record and backward-compute-wait are removed and loss related changes are discarded.

---
In reply to: [477550628](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r477550628) [](ancestors = 477550628,477549484,477541868,477496037,477034463,476905060)